File: solutions/cpp/circular-buffer/1/circular_buffer.cpp
#include "circular_buffer.h"

namespace circular_buffer {

// TODO: add your solution here
template <typename T>
T circular_buffer<T>::read()
{
    if (m_count == 0) throw std::domain_error("cannot read empty buffer");
    T result = m_buffer[m_first];
    m_first = (m_first + 1) % m_len;
    --m_count;
    return result;
}

template <typename T>
void circular_buffer<T>::write(T entry)
{
    if (m_count >= m_len) throw std::domain_error("buffer is full");
    m_buffer[(m_first + m_count) % m_len] = entry;
    ++m_count;
}

template <typename T>
void circular_buffer<T>::clear()
{
    m_count = 0;
}

template <typename T>
void circular_buffer<T>::overwrite(T entry)
{
    m_buffer[(m_first + m_count) % m_len] = entry;
    if (m_count >= m_len) {
        m_first = (m_first + 1) % m_len;
    } else {
        ++m_count;
    }
}

} // namespace circular_buffer

File: solutions/java/circular-buffer/1/src/main/java/CircularBuffer.java
import java.util.Deque;
import java.util.LinkedList;

public class CircularBuffer<T> {
  private final int capacity;
  private final Deque<T> storage = new LinkedList<>();

  public CircularBuffer(int capacity) {
    this.capacity = capacity;
  }

  public T read() throws BufferIOException {
    if (storage.isEmpty()) {
      throw new BufferIOException("Tried to read from empty buffer");
    }

    return storage.pop();
  }

  public void write(T entry) throws BufferIOException {
    if (storage.size() == capacity) {
      throw new BufferIOException("Tried to write to full buffer");
    }

    storage.add(entry);
  }

  public void clear() {
    storage.clear();
  }

  public void overwrite(T entry) {
    try {
      write(entry);
    } catch (BufferIOException e) {
      storage.pop();

      try {
        write(entry);
      } catch (BufferIOException e1) {
        // noop
      }
    }
  }
}
